eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,215 @@
+ * @returns 
+ */
+function arrayToList(inputArray) { 
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+    if (inputArray.length === 0) {
+        return {value: null, rest: null};
+    }
+
+    const maxIndex = inputArray.length - 1;
+
+    function innerArrayToList(currentIndex, list = null) {
+        if (currentIndex < 0) {
+            return list;
+        }
+
+        const arrayElement = inputArray[currentIndex];
+
+        return innerArrayToList(currentIndex - 1, {value: arrayElement, rest: list});
+    }
+
+    return innerArrayToList(maxIndex);
+}
+
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToListIterative(inputArray) {
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+    if (inputArray.length === 0) {
+        return {value: null, rest: null};
+    }
+
+    let listAcc = null;
+
+    for (let i = inputArray.length - 1; i >= 0; i--) {
+        const element = inputArray[i];
+        listAcc = {
+            value: element,
+            rest: listAcc
+        };
+    }
+
+    return listAcc;
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArray(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+
+    function innerListToArray(list, arrayAccumulator = []) {
+        if(list.value !== null) {
+            arrayAccumulator.push(list.value);
+        }
+
+        if (list.rest === null) {
+            return arrayAccumulator;
+        } else {
+            return innerListToArray(list.rest, arrayAccumulator);
+        }
+    }
+
+    return innerListToArray(inputList);
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArrayIterative(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    if (inputList.value === null) {
+        return [];
+    }
+
+    let acc = [];
+
+    let listBuffer = inputList;
+    while (listBuffer !== null) {
+        acc.push(listBuffer.value)
+        listBuffer = listBuffer.rest;
+    }
+
+    return acc;
+}
+
+/**
+ * Returns list from input list with prepended element
+ * @param {*} inputList 
+ * @param {*} element 
+ * @returns 
+ */
+function prepend(inputList, element) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    
+    return {
+        value: element,
+        rest: inputList
+    }
+}
+
+/**
+ * Returns nth element in list
+ * @param {*} inputList 
+ * @param {*} targetIndex 
+ * @returns 
+ */
+function nth(inputList, targetIndex) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    if (isNaN(targetIndex) || targetIndex < 0 || Math.abs(targetIndex) === Infinity) {
+        throw "invalid target index";
+    }
+    if (targetIndex > getListLength(inputList) - 1) {
```

Comment:
This traverses the entire list. You could save this traversal without making the code harder to read by checking for the end of the list in the for loop below.